### PR TITLE
Fix deactivate a user if he does not have a profile

### PR DESCRIPTION
**Pull request by dklimpel**
Fixes: #10240
A user should also be deactivated and mark as erased if he has no profile.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

Signed-off-by: Dirk Klimpel dirk@klimpel.org

### CODE REVIEW

File: synapse/handlers/deactivate_account.py [hunk 1]
```diff
@@ -136,10 +136,19 @@ async def deactivate_account(
...
+                    user, requester, "", by_admin
+                )
+                # Remove displayname from this user
+                await self._profile_handler.set_displayname(
+                    user, requester, "", by_admin
+                )
+            except StoreError as e:
+                # bypass only 404 if user does not have a profile
+                if e.code != 404:
+                    raise e
```
Comment (anoadragon453) [inline synapse/handlers/deactivate_account.py:151]: So the reason a `StoreError` is raised here is due to `set_profile_avatar_url` performing an `update` on `profiles` rather than an `upsert`:

https://github.com/matrix-org/synapse/blob/4b965c862dc66c0da5d3240add70e9b5f0aa720b/synapse/storage/databases/main/profile.py#L83-L91

I think a more fundamental fix for this whole problem would actually just be to convert `set_profile_avatar_url` to use `self.db_pool.simple_upsert`, like so:

```python
        await self.db_pool.simple_upsert(
            table="profiles",
            keyvalues={"user_id": user_localpart},
            values={},
            insertion_values={"avatar_url": new_avatar_url},
            desc="set_profile_avatar_url",
        )
```

it'd probably be a good idea to do the same for `set_profile_displayname` as well, just to avoid the same issue in the future.

File: tests/rest/admin/test_user.py [hunk 2]
```diff
@@ -1098,6 +1098,61 @@ def test_deactivate_user_erase_false(self):
 
         self._is_erased("@user:test", False)
 
+    def test_deactivate_user_erase_true_no_profile(self):
+        """
+        Test deactivating a user and set `erase` to `true`
+        if user has no profile information in database table `profile`
```
Comment (anoadragon453) [inline tests/rest/admin/test_user.py:1104]:
```suggestion
        if user has no profile information (stored in the database table `profiles`).
```

File: tests/rest/admin/test_user.py [hunk 3]
```diff
@@ -1098,6 +1098,61 @@ def test_deactivate_user_erase_false(self):
...
+        )
+
+        self.assertEqual(200, int(channel.result["code"]), msg=channel.result["body"])
+        self.assertEqual("@user:test", channel.json_body["name"])
+        self.assertEqual(False, channel.json_body["deactivated"])
+        self.assertEqual("foo@bar.com", channel.json_body["threepids"][0]["address"])
+        self.assertIsNone(channel.json_body["avatar_url"])
+        self.assertIsNone(channel.json_body["displayname"])
+
+        # Deactivate user
```
Comment (anoadragon453) [inline tests/rest/admin/test_user.py:1128]:
```suggestion
        # Deactivate and erase user
```

File: tests/rest/admin/test_user.py [hunk 4]
```diff
@@ -1098,6 +1098,61 @@ def test_deactivate_user_erase_false(self):
...
+        body = json.dumps({"erase": True})
+
+        channel = self.make_request(
+            "POST",
+            self.url,
+            access_token=self.admin_user_tok,
+            content=body.encode(encoding="utf_8"),
```
Comment (anoadragon453) [inline tests/rest/admin/test_user.py:1135]: You can actually just pass the Python dict directory to `make_request` these days.

File: tests/rest/admin/test_user.py [hunk 5]
```diff
@@ -1098,6 +1098,61 @@ def test_deactivate_user_erase_false(self):
...
+        )
+
+        self.assertEqual(200, int(channel.result["code"]), msg=channel.result["body"])
```
Comment (anoadragon453) [inline tests/rest/admin/test_user.py:1138]:
```suggestion
        self.assertEqual(200, channel.code, msg=channel.json_body)
```

File: tests/rest/admin/test_user.py [hunk 6]
```diff
@@ -1098,6 +1098,61 @@ def test_deactivate_user_erase_false(self):
...
+
+        # Get user
+        channel = self.make_request(
+            "GET",
+            self.url_other_user,
+            access_token=self.admin_user_tok,
+        )
+
+        self.assertEqual(200, int(channel.result["code"]), msg=channel.result["body"])
```
Comment (anoadragon453) [inline tests/rest/admin/test_user.py:1147]:
```suggestion
        self.assertEqual(200, channel.code, msg=channel.json_body)
```

**Comment by anoadragon453** [comment]
Looks like this was introduced in https://github.com/matrix-org/synapse/pull/8932.

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
Thanks for the test! I have a suggestion below for an alternative fix that may prevent other edge cases.

**Comment by dklimpel** [comment]
> Looks like this was introduced in #8932.

Yes. I didn't realise that there were users without profiles.

File: synapse/handlers/deactivate_account.py (hunk 1 above)
Comment (dklimpel) [reply synapse/handlers/deactivate_account.py:151]: Ok. I will make an update.

**Comment by richvdh** [comment]
why *do* we have users without profiles?

**Comment by dklimpel** [comment]
> why _do_ we have users without profiles?

If you create a user without displayname
https://github.com/matrix-org/synapse/blob/f0e02f5df2bc3ae779ac8c18578deebdfecc7e97/synapse/storage/databases/main/registration.py#L1563-L1575

no profile will be created
https://github.com/matrix-org/synapse/blob/f0e02f5df2bc3ae779ac8c18578deebdfecc7e97/synapse/storage/databases/main/registration.py#L1629-L1638

**Comment by dklimpel** [comment]
> why _do_ we have users without profiles?

https://github.com/matrix-org/synapse/issues/10240#issuecomment-868986543

**Comment by richvdh** [comment]
> > why _do_ we have users without profiles?
> 
> If you create a user without displayname ... no profile will be created

ugh.

File: tests/rest/admin/test_user.py [hunk 7]
```diff
@@ -1098,6 +1094,59 @@ def test_deactivate_user_erase_false(self):
 
         self._is_erased("@user:test", False)
 
+    def test_deactivate_user_erase_true_no_profile(self):
+        """
+        Test deactivating a user and set `erase` to `true`
+        if user has no profile information (stored in the database table `profiles`).
+        """
+
+        # Delete profile information
```
Comment (richvdh) [inline tests/rest/admin/test_user.py:1103]:
```suggestion
        # Users normally have an entry in `profiles`, but occasionally they are created without one.
        # To test deactivation for users without a profile, we delete the profile information for our user.
```

**Review by richvdh — APPROVED** [review]
lgtm, thanks!
